DOCSP-31502: regex topic namespace mapping (#141)
mongodb/docs-kafka-connector: source/source-connector/usage-examples/topic-naming.txt

Once set, your connector publishes any changes to the ``data`` collection
in the ``test`` database to the Kafka topic named ``test.data.mySuffix``.

.. _topic-naming-namespace-map:

Topic Namespace Map
-------------------

You can configure your source connector to map namespace values to Kafka
topic names for incoming change event data. Topic namespace maps contain
**pairs** that are made up of a namespace pattern and a destination

> topic.

Reviewer: > and a destination topic

That's not a topic, but a topic name template. The topic name is then computed based on the namespace that matched the namespace pattern and the corresponding topic name template (note how the text proposed in DOCSP-31502 uses the terms "topic name" and "topic name template").

There are other places where the documentation does not differentiate these two terms, for example:

- _"The following example shows how to specify the `topic.namespace.map` setting to define a topic namespace mappings from the `carDb` database to the `automobiles` topic and the `carDb.ev` namespace to the `electricVehicles` topic: `topic.namespace.map={"carDb": "automobiles", "carDb.ev": "electricVehicles"}`"_. Here `automobiles` is a topic name template, not a topic name. This becomes clear when we read the example further and it says _"If the change event came from the database `carDb` and a collection other than `ev`, the connector sets the destination to the `automobiles.<collection name>` topic."_: the computed topic name is not `automobiles`, but `automobiles.<collection name>`, because `automobiles` is a topic name template. `electricVehicles` obviously must be conceptually the same thing as `automobiles`, i.e., a topic name template. It just so happens that the `electricVehicles` topic name template does not have anything to expand/add (barring the `topic.prefix`/`topic.suffix`), and matches the topic name computed based on it.
- https://github.com/mongodb/docs-kafka-connector/pull/141#discussion_r1346394167

Author: Thanks for this explanation, I definitely see your point and understand. Let me go through the page and correct the terminology where needed.

  To learn more about this property, see :ref:`source-configuration-kafka-topic`.
- ``coll``: The collection name from the matching namespace, or an empty
  string if there is no collection name.
- ``sep_coll``: The value of the ``coll`` variable prefixed with the value
  of the ``sep`` variable, or an empty string if the value of ``coll`` is empty.
- ``coll_sep``: The value of the ``coll`` variable suffixed with the
  value of the ``sep`` variable, or an empty string if the value of ``coll`` is empty.
- ``sep_coll_sep``: The value of the ``coll`` variable prefixed and
  suffixed with the value of the ``sep`` variable, or an empty string if
  the value of ``coll`` is empty.
 
If you use any of the previous variables in your topic name template,
you must enclose them in curly brackets (``{}``) for the connector to
expand them. You can't use curly brackets in the topic name template for
any other purpose.

.. note:: Escaping Characters

   If your namespace pattern includes characters that need escaping,
   you must prefix these characters with two backslashes (``\\``) in the
   pattern. For example, to match a period (``'.'``) character, you must

>    write it as ``'\\.'``.

Author: Reverted the note to preserve the original level of detail

>    write it as ``'\\.'``.

Reviewer: > However, you must escape the backslash ``\`` character as ``\\``.

The original text mentioned that this second escaping is due to the JSON syntax:

```
However, reverse solidus '\' itself must be escaped as '\\' according to the JSON syntax.
```

What is this part remains removed?

>    write it as ``'\\.'``.

Author: Will add this sentence

 
The connector matches namespaces in the following order:
 
1. Pairs with database and collection names in the namespace pattern. To
   learn more about this namespace pattern, see the :ref:`Topic
   Namespace Map Example <topic-naming-namespace-map-example>`.
#. Pairs with only a database name in the namespace pattern. To
   learn more about this namespace pattern, see the :ref:`Topic
   Namespace Map Example <topic-naming-namespace-map-example>`.

> #. Regex pairs in order.

Author: added detail and moved content up

#. The wildcard pair. To learn more about this namespace pattern, see
   the :ref:`Topic Namespace Map with Wildcard Example

>    <topic-naming-namespace-map-wildcard-example>`.

Author: I can move this to the configuration property description instead.

>    <topic-naming-namespace-map-wildcard-example>`.

Reviewer: 👍

This example shows how to specify the ``topic.namespace.map``
setting so that the connector performs the following mapping:

- Writes events from databases that match a regular expression to the

>   ``industrial{sep_coll}`` topic. The regex pattern matches any database

Author: I did not mention the value of `topic.separator` because `.` is the default, can fix.
I don't understand the contradiction you pointed out.

I thought if the collection name was `something`, then {sep_coll} would be expanded to `.something` (default separator). Is this not what I described?

In any case, I am going to use the example from the ticket instead, as you requested in your next comment.

>   ``industrial{sep_coll}`` topic. The regex pattern matches any database

Reviewer: > I thought if the collection name was something, then {sep_coll} would be expanded to .something (default separator).

Exactly, which means that `industrial{sep_coll}` is not a topic name, as it is stated in the docs at the moment. But because the distinction between a topic name template and a topic name is not done in the docs, we had the following two lines in the PR when I this thread was started:

> ... to the ``industrial{sep_coll}`` topic.
> ... to the ``industrial.<collectionName>`` topic.

The first one is a topic name template, the second one is, indeed, a topic name<sup>1</sup>. If the documentation makes the distinction, one will be able to see why the text in those documentation lines is different. I tried to expand more on this in https://github.com/mongodb/docs-kafka-connector/pull/141#discussion_r1349922059, where a similar issue exists in a couple of different places. If I failed to get a point across again, I'll not push this issue further in either of these threads.

---

<sup>1</sup> Strictly speaking, one may argue that `industrial.<collectionName>` is also template, and in a way it is: it shows what is computed based on the topic name template `industrial{sep_coll}` without focusing on a specific collection name. But even if so, `industrial.<collectionName>` is not the same thing as a template from pairs in `topic.namespace.map`.

  names that include the string ``'vertical'``, even if the database

>   name has other characters.

Reviewer: The regex does not match namespaces with database names that are anything but `vertical` exactly. The original text describes what the regex matches as follows (replace `myDb` with `vertical`):

```
The regex namespace pattern matches any namespace with the 'myDb' database name.
```

Reviewer: I did not review the examples and their descriptions further. Is it an option to take the examples from the originally proposed text, replace `myDb`, `myColl`, `topicOne`, ... with something meaningful, if this deemed necessary, but otherwise leave the structure, the descriptions and explanations virtually unchanged? Those examples have already been analyzed, tested, and their descriptions are much more likely to be correct than new descriptions written for new tests, or even for the same tests.

Author: thanks for pointing it out that I did not interpret the regex correctly. I will use the example from the ticket instead and just change the names. I thought that I had mostly done that in this draft, with the exception of leaving out the wildcard as it is described in the next example.

- If the change document came from any database other than ``carDb``, the
  connector sets the destination topic to the default namespace naming
  scheme.
- If you define the ``topic.prefix`` and ``topic.suffix`` settings, the
  connector applies their values to the destination topic name after it
  performs the namespace mapping.

.. _topic-naming-namespace-map-regex-example:

Topic Namespace Map with Regular Expressions Example
----------------------------------------------------

> You can use regular expressions (regex) to specify namespaces within a

Reviewer 2: S:

```suggestion
You can use regular expressions (regexes) to specify namespaces within a
```

topic name template. The connector supports the following variables: 
 
- ``db``: The database name from the matching namespace.
- ``sep``: The value of the ``topic.separator`` configuration property.
  To learn more about this property, see :ref:`source-configuration-kafka-topic`.
- ``coll``: The collection name from the matching namespace, or an empty
  string if there is no collection name.
- ``sep_coll``: The value of the ``coll`` variable prefixed with the value
  of the ``sep`` variable, or an empty string if the value of ``coll`` is empty.
- ``coll_sep``: The value of the ``coll`` variable suffixed with the
  value of the ``sep`` variable, or an empty string if the value of ``coll`` is empty.
- ``sep_coll_sep``: The value of the ``coll`` variable prefixed and
  suffixed with the value of the ``sep`` variable, or an empty string if
  the value of ``coll`` is empty.
 
If you use any of the previous variables in your topic name template,
you must enclose them in curly brackets (``{}``) for the connector to
expand them. You can't use curly brackets in the topic name template for
any other purpose.

.. note:: Escaping Characters

>    You must follow the JSON syntax to create a namespace pattern that includes
>    characters that need escaping. For example, to match
>    ``'.'``, the regex syntax requires that you escape it as ``'\.'``
>    You must escape the backslash ``'\'`` character as ``'\\'``.
>    Consequently, to match ``'.'`` in your namespace pattern, you must
>    write it as ``'\\.'``.

Reviewer 2: ```suggestion
   If your namespace pattern includes characters that must be escaped, you must prefix
   these characters with two backslashes (``\\``). For example, to match a period (``.``),
   you must write it as ``\\.``.
```

Reviewer 2: just a draft--feel free to edit!

  ``topic.suffix`` settings to the destination topic name after it
  performs namespace mapping.

.. _topic-naming-namespace-map-regex-example:

> Topic Namespace Map with Regex Example

Author: Follow up: do you think all these headers need the word example?

> Topic Namespace Map with Regex Example

Reviewer 2: Nope! I think I'd remove them.

--------------------------------------

You can use regular expressions as namespace patterns when specifying a
topic namespace map. To specify a regular expression in a namespace, start the
namespace pattern with the solidus (``/``) character. Create the regular
expression following the syntax and behavior outlined by the
``java.util.regex.Pattern`` class.

.. important::
   
   MongoDB database names cannot contain the ``/`` character. If your
   namespace contains this character, the connector interprets the
   namespace as a regex pattern or raises a ``ConnectConfigException``

>    if the ``/`` character is not at the beginning of the namespace.

Author: hope the reword clarifies things!

>    if the ``/`` character is not at the beginning of the namespace.

Reviewer 2: perfect

The connector computes the topic name by doing variable expansion on the
topic name template. The connector supports the following variables: 
 
- ``db``: The database name from the matching namespace.
- ``sep``: The value of the ``topic.separator`` configuration property.
  To learn more about this property, see :ref:`source-configuration-kafka-topic`.
- ``coll``: The collection name from the matching namespace, or an empty
  string if there is no collection name.
- ``sep_coll``: The value of the ``coll`` variable prefixed with the value

>   of the ``sep`` variable if the value of ``coll`` is not empty.

Author: nothing at all - the placeholder would be empty. I can clarify this in the list items that it applies to.

- ``coll_sep``: The value of the ``coll`` variable suffixed with the
  value of the ``sep`` variable if the value of ``coll`` is not empty.
- ``sep_coll_sep``: The value of the ``coll`` variable prefixed and
  suffixed with the value of the ``sep`` variable if the value of
  ``coll`` is not empty.
 
You must enclose variables between curly brackets (``{}``) for the connector to
expand them. You can't use curly brackets in the topic name template for

> any other purpose.

Author: I would rather avoid stacking admomitions here since the escaping characters note is right below.

.. note:: Escaping Characters

   You must follow the JSON syntax to create a namespace pattern that includes
   characters that need escaping. For example, to match
   ``'.'``, the regex syntax requires that you escape it as ``'\.'``

>    You must escape the reverse solidus ``'\'`` character as ``'\\'``.

Author: Yep! thats how i understood it. https://stackoverflow.com/questions/17711831/double-backslash-on-quotes-when-json-stringifying-an-array-of-strings

   Consequently, to match ``'.'`` in your namespace pattern, you must
   write it as ``'\\.'``.
 
The connector matches namespaces by using the following order:
 
1. Pairs with database and collection names in the namespace pattern. To
   learn more about this namespace pattern, see the :ref:`Topic
   Namespace Map Example <topic-naming-namespace-map-example>`.
#. Pairs with only a database name in the namespace pattern. To
   learn more about this namespace pattern, see the :ref:`Topic
   Namespace Map Example <topic-naming-namespace-map-example>`.
#. Regex pairs in order.
#. The wildcard pair. To learn more about this namespace pattern, see
   the :ref:`Topic Namespace Map with Wildcard Example
   <topic-naming-namespace-map-wildcard-example>`.

This example shows how to specify the ``topic.namespace.map``
setting to define the following topic namespace mapping:

- Maps a regular expression to the topic ``industrial{sep_coll}``. The
  regex pattern matches any database names that include the string
  ``'vertical'``, even if the database name has other characters.

> - Maps the ``vertical.health`` namespace to the ``healthcare`` topic.

Reviewer 2: I: This information is good, but the format is a little confusing. Starting the list items with "Maps" implies a subject (the example?) that isn't clear from the lead-in clause.

Author: I will revise so that this is more clear.

.. code-block:: ini

   topic.namespace.map={"/vertical(?:\\..*)?": "industrial{sep_coll}", "vertical.health": "healthcare"}

> Since the ``vertical.health`` namespace mapping takes precedence over

Reviewer 2: I: [Since](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#s:~:text=press%20Enter.-,since,-Use%20since%20to)

```suggestion
Because the ``vertical.health`` namespace mapping takes precedence over
```

the regular expression namespace mapping, the connector performs the
following actions:

- If the change event comes from the database ``vertical`` and
  collection ``health``, the connector sets the destination to the
  ``healthcare`` topic.

> - If the change event comes from any database that gets
>   matched by the regex, the connector sets the destination to the

Reviewer 2: ```suggestion
- If the change event comes from any database that
  matches the regex, the connector sets the destination to the
```

  ``industrial.<collectionName>`` topic. The following examples
  demonstrate this mapping:

  - If the change event comes from the ``vertical.wasteManagement``
    namespace, the connector writes to the
    ``industrial.wasteManagement`` topic.

  - If the change event comes from the ``vertical_auto.batteries``
    namespace, the connector writes to the
    ``industrial.batteries`` topic because the database name contains
    the string ``'vertical'``.
  
  - If the change event comes from the ``vertical_auto``
    database but no specific collection, the connector writes to the
    ``industrial`` topic.

> - If the change document comes from any database that does not get
>   matched by the regex, the connector sets the destination topic to the

Reviewer 2: ```suggestion
- If the change document comes from any database that doesn't
  match the regex, the connector sets the destination topic to the
```

  default namespace naming scheme. 

> - If defined, the connector applies the ``topic.prefix`` and
>   ``topic.suffix`` settings to the destination topic name after it
>   performs namespace mapping.

Reviewer 2: I: I think this means if these settings are defined, but as written, it means "If the connector is defined..."

```suggestion
- If the ``topic.prefix`` and ``topic.suffix`` settings are defined, the connector applies
  their values  to the destination topic name after it
  performs namespace mapping.
```

   When you create a namespace pattern, ensure that you properly handle
   characters that need escaping. For example, to match
   ``.``, the regex syntax requires that you escape it as ``\.``
   However, you must escape the backslash ``\`` character as ``\\``
   according to the JSON syntax. Then, to match ``.`` in your namespace
   pattern, you must write it as ``\\.``.

This example shows how to specify the ``topic.namespace.map``
setting so that the connector performs the following mapping:

- Writes events from databases that match a regular expression to the

>   ``industrial{sep_coll}`` topic name template. The regex pattern matches any

Reviewer: Replacing "topic" with "topic name template" where appropriate without changing the surrounding text results weird text, for example:

> Writes events from databases that match a regular expression to the `industrial{sep_coll}` topic name template.

Events cannot be written to a topic name template, but they can be written to a topic that was determined based on the topic name template. I am pointing this out just in case you want to work on the wording, I don't see it as part of the technical review.
